### ISSUE TITLE
Consider using go-prompt to enable a nice interactive prompt like mssql-cli

### ISSUE THREAD

**Issue by daviewales** [microsoft/go-sqlcmd#481]
mssql-cli has a nice prompt and nice table output thanks to the use of the following Python libraries:

- [CLI Helpers](https://cli-helpers.readthedocs.io/en/stable/)
- [Python Prompt Toolkit](https://python-prompt-toolkit.readthedocs.io/en/master/)
- [Tabulate](https://github.com/astanin/python-tabulate)

mssql-cli is on the deprecation track, and suggests that go-sqlcmd is the future.
It would be nice if go-sqlcmd copied some of the nice usability features from mssql-cli.

There are existing Go libraries inspired by Python Prompt Toolkit and Tabulate which may help improve the go-sqlcmd interface similar to mssql-cli:

Tabulate inspired libraries:

- https://github.com/markkurossi/tabulate
- https://github.com/rgeoghegan/tabulate
- https://github.com/bndr/gotabulate

Non-tabulate-inspired table library:

- https://github.com/jedib0t/go-pretty

Python Prompt Toolkit inspired library:

- https://github.com/c-bata/go-prompt

Would go-sqlcmd consider implementing some of these usability features?

**Comment by michaelstack** [comment]
I would also like to see go-sqlcmd implement more of the functionality that is available in mssql-cli and the other tools in the dbcli family (https://github.com/dbcli).

- Tab completion of table names and other object names
- show long output in a user-configurable pager
- ability to transpose records like postgres psql's `\x` command
- in interactive sessions, have a setting to invoke a statement without having to put `GO` on a newline

**Comment by michaelstack** [comment]
@shueybubbles 

> Regarding `transpose records` is that similar to `sqlcmd -F vertical` ?

Yes, and it is also like these:

```
:setvar SQLCMDFORMAT vertical
:setvar SQLCMDFORMAT horizontal
```

The advantage of psql's `\x` command is that it is much easier to type. Plus the one single command toggles the setting, as opposed to two separate commands.

Psql also has `\x auto`, which will automatically switch based on the width of the records selected.

This automatic behaviour is included in pgcli under the `auto_expand` setting:

https://github.com/dbcli/pgcli/blob/96eb37fd1989ec8fd86150f20053abd03f50e7a4/pgcli/pgclirc#L51

Looks like it was included in mssql-cli also with the `expand` setting:

https://github.com/dbcli/mssql-cli/blob/67a3a4364383d11776bb0a397ceae7bc19200e41/mssqlcli/mssqlclirc#L29

**Comment by pietdaniel** [comment]
+1

**Comment by azinsharaf** [comment]
+1

**Comment by AnhQuanTrl** [comment]
I think this is not ready to replace mssql-cli yet if tab-completion is not supported. This capabilities make using DB CLI tool a comparable alternative to full-fleshed DB tool such as DBeaver.

**Comment by azinsharaf** [comment]
i checked usql cli and it had some tab completion capabilities. i haven't spent enough time to check it throughly.

**Comment by rsmck** [comment]
Agreed, there's a brilliant animation on the mysql-cli home page boasting these excellent features; 

- Auto-completion: fewer keystrokes needed to complete complicated queries.
- Syntax highlighting: highlights T-SQL keywords.
- Query history: easily complete an auto-suggested query that was previously executed.
- Configuration file support: customize the mssql-cli experience for your needs.
- Multi-line queries: execute multiple queries at once using the multi-line edit mode.
- Non-interactive support: execute a query without jumping into the interactive experience.

Directly below a notice saying that this project replaces it ... which lacks all of them. 

Feels like a step backwards.
